src/node_contextify.cc

```diff
@@ -286,6 +286,14 @@ void ContextifyContext::WeakCallback(
   delete context;
 }
 
+void ContextifyContext::WeakCallbackCompileFn(
+    const WeakCallbackInfo<CompileFnEntry>& data) {
+  CompileFnEntry* entry = data.GetParameter();
+  entry->env->id_to_function_map[entry->id].Reset();
```

Comment:
Can't we just use `node::Persistent` here?

Reply:
Can you clarify what you mean in the code?

Reply:
@guybedford The suggestion is to replace `v8::Persistent` with `node::Persistent` (aka just `Persistent` in most of our code), because that one calls `Reset()` automatically when it is destroyed.

Reply:
Btw, this does get interesting for `CompileFnEntry` objects that still exist when the `Environment` is torn down. We never delete these objects, and we don’t call the weak callbacks either. This definitely still leaks memory. :/

Reply:
Ahh thanks so much for explaining, that makes sense. I've switched to the Node persistent now, please review. Also if you have a code suggestion for the Reset emplace issue that would be welcome too.

Reply:
@guybedford Sorry, this still leaks memory for `CompileFnEntry`s that stick around longer than the `Environment` (which I assume is common). That could be addressed in multiple ways:

- @joyeecheung’s suggestion of wrapping it in a `BaseObject`, also making it heap-snapshot-trackable – this might come with some overhead, and probably requires some fiddling getting it to attach to the function object in the right way
- Manually deleting from the `Environment` destructor – this kind of breaks encapsulation, and it requires a way for us to access all existing `CompileFnEntry` from there?

Here are my steps to reproduce/verify the memory leak, if that helps:

```console
# checkout this PR
$ git fetch upstream && git checkout upstream/pr/21573
$ git rebase upstream/master
# compile
$ make -j4
# run test + see memory leak
$ valgrind --leak-check=full --show-leak-kinds=definite ./node test/parallel/test-worker-stdio.js 
[...]
==1868==    at 0x4C3045A: operator new(unsigned long) (vg_replace_malloc.c:344)
==1868==    by 0x797DA1: node::contextify::ContextifyContext::CompileFunction(v8::FunctionCallbackInfo<v8::Value> const&) (in /home/sqrt/src/node/master/out/Release/node)
[...]
```

Reply:
Hmm, I thought the point of using node::Persistent is that it would be automatically disposed with the environment?

Reply:
The point is pretty much just that it `Reset()`s automatically – which V8’s `Persistent` should have done from the beginning. (I assume they didn’t change it for backwards compatibility.)

We don’t keep track of all `Persistent` handles and which `Environment` – if any – they are associated with, so doing that would be tricky, too.

Reply:
(Btw, I think I could come up with an example patch for the `BaseObject` suggestion, if that would help you.)

Reply:
I was just looking at implementing the second option of maintaining a set of the CompileFnEntry items on the env record to dispose.

If you think the BaseObject approach is preferable though that works too.

Reply:
You can feel free to do that, sure. :)

Reply:
Ok, I've implemented this in the latest commit. Review would be appreciated :)

Also to note again - this entire architecture should be changed soon, in case that isn't completely obvious.